cygwin/Platform.c

```diff
@@ -0,0 +1,339 @@
+      v[CPU_METER_KERNEL]  = cpuData->systemPeriod / total * 100.0;
+      v[CPU_METER_IRQ]     = cpuData->irqPeriod / total * 100.0;
+      v[CPU_METER_SOFTIRQ] = cpuData->softIrqPeriod / total * 100.0;
+      this->curItems = 5;
+
+      v[CPU_METER_STEAL]   = cpuData->stealPeriod / total * 100.0;
+      v[CPU_METER_GUEST]   = cpuData->guestPeriod / total * 100.0;
+      if (settings->accountGuestInCPUMeter) {
+         this->curItems = 7;
+      }
+
+      v[CPU_METER_IOWAIT]  = cpuData->ioWaitPeriod / total * 100.0;
+   } else {
+      v[CPU_METER_KERNEL] = cpuData->systemAllPeriod / total * 100.0;
+      v[CPU_METER_IRQ] = (cpuData->stealPeriod + cpuData->guestPeriod) / total * 100.0;
+      this->curItems = 4;
+   }
+
+   percent = sumPositiveValues(v, this->curItems);
+   percent = MINIMUM(percent, 100.0);
+
+   if (settings->detailedCPUTime) {
+      this->curItems = 8;
+   }
+
+   v[CPU_METER_FREQUENCY] = cpuData->frequency;
+
+   v[CPU_METER_TEMPERATURE] = NAN;
+
+   return percent;
+}
+
+void Platform_setMemoryValues(Meter* this) {
+   const Machine* host = this->host;
+   long int usedMem = host->usedMem;
+   long int buffersMem = host->buffersMem;
+   long int cachedMem = host->cachedMem;
+   usedMem -= buffersMem + cachedMem;
+   this->total = host->totalMem;
+   this->values[MEMORY_METER_USED] = usedMem;
+   this->values[MEMORY_METER_BUFFERS] = buffersMem;
+   // this->values[MEMORY_METER_SHARED] = "shared memory, like tmpfs and shm"
+   // this->values[MEMORY_METER_COMPRESSED] = "compressed memory, like zswap on linux"
+   this->values[MEMORY_METER_CACHE] = cachedMem;
+   // this->values[MEMORY_METER_AVAILABLE] = "available memory"
+
+}
+
+void Platform_setSwapValues(Meter* this) {
+   const Machine* host = this->host;
+   this->total = host->totalSwap;
+   this->values[SWAP_METER_USED] = host->usedSwap;
+   // this->values[SWAP_METER_CACHE] = "pages that are both in swap and RAM, like SwapCached on linux"
+   // this->values[SWAP_METER_FRONTSWAP] = "pages that are accounted to swap but stored elsewhere, like frontswap on linux"
+}
+
+char* Platform_getProcessEnv(pid_t pid) {
+   char procname[128];
+   xSnprintf(procname, sizeof(procname), PROCDIR "/%d/environ", pid);
+   FILE* fd = fopen(procname, "r");
+   if (!fd)
+      return NULL;
+
+   char* env = NULL;
+
+   size_t capacity = 0;
+   size_t size = 0;
+   ssize_t bytes = 0;
+
+   do {
+      size += bytes;
+      capacity += 4096;
+      env = xRealloc(env, capacity);
+   } while ((bytes = fread(env + size, 1, capacity - size, fd)) > 0);
+
+   fclose(fd);
+
+   if (bytes < 0) {
+      free(env);
+      return NULL;
+   }
+
+   size += bytes;
+
+   env = xRealloc(env, size + 2);
+
+   env[size] = '\0';
+   env[size + 1] = '\0';
+
+   return env;
+}
+
+FileLocks_ProcessData* Platform_getProcessLocks(pid_t pid) {
+   // TODO
+   (void)pid;
+   return NULL;
+}
+
+void Platform_getFileDescriptors(double* used, double* max) {
+   // TODO
+   *used = NAN;
+   *max = NAN;
```

Comment:
> TL;DR: 16,711,680 on x64 and 16,744,448 on x86; per process.

but this is a global limit of fds, we may need to multiply it by the maximum number of processes.

also, does it apply to cygwin? some fds can be wrapped by cygwin and are not actual win32 handles.

Reply:
Those limits are purely based on the implementation in the WinAPI and thus cygwin somewhat inherits them. But actual limits in cygwin may be lower or higher (e.g. when cygwin shares one handle for a resource). And yes, those numbers would need to be multiplied by the number of processes.

Though in all fairness: For the FD meter to be somewhat sensible, we can assume those as global limits and remember the max open if that was ever above the number of opened handles. Because we know at least that many handles can exist and tracking the actual number will be rather pointless for the actual display in the bar meter … (NB: You can report different numbers for the text representation and the bar graph).

@fasterit IIRC there was Solaris where things where more or less similarly dynamic?

Reply:
> @fasterit IIRC there was Solaris where things where more or less similarly dynamic?

Yes, there is no global limit (RAM is the limit). But a hard and a soft limit per user and per process.
Which also change between minor versions. Because $reasons (https://support.oracle.com/knowledge/Sun%20Microsystems/2736930_1.html [paywalled]).
But the good news is ... none of that is implemented to be visualized in htop for Solaris :-).